server/etcdserver/txn/txn.go

```diff
@@ -235,8 +243,31 @@ func Txn(ctx context.Context, lg *zap.Logger, rt *pb.TxnRequest, txnModeWriteWit
 	} else {
 		txnWrite = mvcc.NewReadOnlyTxnWrite(kv.Read(mvcc.ConcurrentReadTxMode, trace))
 	}
+	txnPath, err := txnCompare(ctx, txnWrite, rt, isWrite, lessor)
+	if err != nil {
+		txnWrite.End()
+		return nil, nil, err
+	}
+	// When executing mutable txnWrite ops, etcd must hold the txnWrite lock so
+	// readers do not see any intermediate results. Since writes are
+	// serialized on the raft loop, the revision in the read view will
+	// be the revision of the write txnWrite.
+	if isWrite {
+		txnWrite.End()
+		txnWrite = kv.Write(trace)
+	}
+	txnResp, err := txn(ctx, lg, txnWrite, rt, isWrite, txnPath)
```

Comment:
Originally, (If I understand correctly) there are two phases to check lease ID is valid: `checkRequests`(compare txn) and `applyTxn -> Put`. In this change, we lost the lease ID checker in `applyTxn`. Since we open new transaction after `txnCompare`, it might cause race here? Just have one concern about this.

Reply:
Looks like the `applyTxn -> put` always skips the step to check if the lease ID exists or not before or after the change because the write transaction is already opened after right after `checkRequests`. 

https://github.com/etcd-io/etcd/blob/9e1abbab6e4ba2886238f49b0d48fc19a546c7cf/server/etcdserver/txn/txn.go#L47-L55

So the race could be always there theoretically. The argument of the race does not exist is associating the nonexist lease could fail in [storeTxnWrite.put](https://github.com/etcd-io/etcd/blob/29769984e62ec9422ff0555ada919d913549d7c1/server/storage/mvcc/kvstore_txn.go#L258)

Reply:
oh, yeah. There is if-condition about the transaction.